assets/shiny-server.js

```diff
@@ -273,10 +317,85 @@
           debug("NOT opening channel " + channel.id);
         }
       }
+      
+      // Send any buffered messages.
+      var msg;
+      while ((msg = self._buffer.shift())){
+        self._conn.send(msg);
+      }
     };
-    this._conn.onclose = function(e) {
+
+    // @param reconnect If true, show the reconnecting dialog. If false, show 
+    // disconnected
+    this.startDisconnect = function(){
+      self._disconnectDialog = true;
+      $('body').addClass('reconnecting');
+
+      var timeout = 15;
+      var reconnectingContent = '<button type="button" id="ss-reconnect-btn" class="ss-dialog-button">Reconnect ('+timeout+')</button><span class="ss-dialog-text">Trouble connecting to server</span>';
+      $('<div id="ss-connect-dialog">'+reconnectingContent+'<div class="ss-clearfix"></div></div><div id="ss-gray-out"></div>').appendTo('body');
+      $('#ss-reconnect-btn').click(function(){
+        $('ss-reconnect-btn').prop('disabled', true);
+        log("Attempting to reopen.");
+        self._openConnection();
+      });
+
+      var countdown = setInterval(function(){
```

Comment:
I think it's important to communicate when we're going to transition from "reconnect your existing session" to "start a new session."

Either way, this will go away when we auto-reconnect.